feat: Kramer switches PPT-367

pkheav opened the pull request:
I've ported 2/3 [drivers from here](https://github.com/acaprojects/ruby-engine-drivers/tree/beta/modules/kramer/switcher) but not [here](https://github.com/acaprojects/ruby-engine-drivers/blob/beta/modules/kramer/switcher/presentation.rb) as it is just a simple extension of the driver p3000 driver. 

I can port the presentation.rb driver if needed though as it is simple.

Closes #47

stakach reviewed (changes requested):
These drivers should implement the [Switchable](https://github.com/PlaceOS/driver/blob/master/src/placeos-driver/interface/switchable.cr#L9) interface

@kimburgess we should probably update your [Extron Matrix driver](https://github.com/PlaceOS/drivers/blob/master/drivers/extron/matrix.cr) to do the same

The switchable interface was brainstormed early on, so happy for updates to made to the interface too.

camreeves committed:
chore(leviton): add debugging to acquisuite driver (#381)

* chore(leviton): add debugging to acquisuite driver

* chore(leviton): dig debugging lines

Co-authored-by: Stephen von Takach <steve@place.technology>

* chore(leviton): add better debugging to acquisuite driver

Co-authored-by: Stephen von Takach <steve@place.technology>

stakach committed:
fix(sony\camera\visca): zoom position on preset recall

stakach committed:
feat(aver/cam520_pro): add aver cam520 driver

stakach committed:
chore(aver/cam520_pro): add additional logging

stakach committed:
fix(aver/cam520_pro): add before_request for setting auth header

stakach committed:
chore(aver/cam520_pro): log the request path

stakach committed:
fix(aver/cam520_pro): fix hostname when not running a spec

w-le committed:
feat(infosilem/mock): add mock driver for testing RoomSchedule

w-le committed:
fix(infosilem/mock): args for bookings func

w-le committed:
fix(infosilem/mock): bookings? response

camreeves committed:
fix(leviton): small webhook fixes (#382)

* fix(leviton): final logic fixes

* chore(leviton): crystal tool format

stakach committed:
fix(aver/cam520_pro): http requests shouldn't use the websocket port

stakach committed:
fix(aver/cam520_pro): save presets

stakach committed:
fix(aver/cam520_pro): any stop command will stop movement

stakach committed:
Revert "fix(aver/cam520_pro): any stop command will stop movement"

This reverts commit 3fe34ecef191f0b6b6102f0be5899604d4c1f5ca.

stakach committed:
chore(aver/cam520_pro): debug body data

stakach committed:
fix(aver/cam520_pro): compare absolute values when moving

stakach committed:
feat(crestron/nvx): add virtual switcher

grkek committed:
Test merge for Cisco ISE

stakach committed:
fix(aver/cam520_pro): ensure presets are exposed on load

add dummy powerable interface

stakach committed:
fix: camera presets status

stakach committed:
feat(place/http_pinger): add curl helper

also adds details logging on requests

stakach committed:
fix(event_attendance_recorder): ensure no concurrent state updates

stakach committed:
feat: add kramer RC8 touch pad driver

stakach committed:
chore(event_attendance_recorder): remove old state log output

stakach committed:
fix(place\calendar_common): prevent rate limiter error

stakach committed:
fix(place\calendar_common): prevent rate limiter error

stakach committed:
chore(juniper\mist_websocket): update default WS API

stakach committed:
feat(juniper\mist_websocket): add some debugging logs

stakach committed:
fix(juniper\mist_websocket): different host for REST API requests

stakach committed:
chore(juniper/mist): fix spec

stakach committed:
fix(juniper\mist_models): mark optional map fields

stakach committed:
fix(juniper\mist_models): mark optional client fields

stakach committed:
fix(juniper\mist_models): fix accuracy

stakach committed:
feat(gallagher\rest_api): implement DoorSecurity interface (#386)

stakach committed:
fix(gallagher\rest_api): use update URL in responses

stakach committed:
fix(gallagher\rest_api): may not be using PDFs

stakach committed:
fix(gallagher\rest_api): ids are strings

stakach committed:
chore(rhombus\security_interop): add some debugging output

stakach committed:
feat(gallagher\rest_api): add some helpful requests

stakach committed:
feat(interface/door_security): allow for custom channel

defaults to the original channel

stakach committed:
feat(place/visitor_mailer): add option for sending reminder emails

stakach committed:
fix(place/staff_api): zones parents request

stakach committed:
chore(shard.yml): bump version

stakach committed:
fix(juniper/mist_websocket): data is not always provided

stakach committed:
feat: add room control sensor shutdown helper

stakach committed:
fix(crestron/occupancy_sensor): expose a sensor interface compatible state

grkek committed:
feat(cisco/ise): support ISE REST API v3.1

* Reverted back to using PlaceOS proxy client and implement the request logic in the driver

* Update the type of custom attributes

* Add default values for the JSON model

* Add a reminder

* Add lookup by email and password update logic

* chore(cisco/ise): default timezone, error msg, comments

Co-authored-by: William Le <will@place.technology>

w-le committed:
fix(placeos/staff-api): query_guests

w-le committed:
fix(visitor_mailer): time format for query_guests

w-le committed:
fix(placeos/visitor_mailer): zones param of query_guests

w-le committed:
fix(placeos/visitor_mailer): building zone var

w-le committed:
fix(placeos/visitor_mailer): building_zone safe accessor

w-le committed:
fix(placeos/visitor_mailer): building_zone var ref

grkek committed:
fix(Cisco/ise): Exclude optional fields from the Internal User model (#400)

* Update from XML to JSON using the same Authorization technique

* run formatting and update the internal user model to exclude unused fields

grkek committed:
fix(cisco/ise): remove custom attributes from ISE (#401)

* Update from XML to JSON using the same Authorization technique

* run formatting and update the internal user model to exclude unused fields

* chore: remove custom attributes for base functionality

stakach committed:
chore(place/bookings): add function to get calendar id

stakach committed:
feat(cisco/room_kit): implement sensor interface

stakach committed:
fix(place/booking_notifier): don't notify unless approved

and process_state is not notified

stakach committed:
fix(place/booking_notifier): only send email once booking is approved

stakach committed:
feat(place/booking_notifier): send cancelled emails

stakach committed:
chore(cisco/room_kit): add useful defaults

stakach committed:
fix(event_attendance_recorder): ensure no negative values

w-le committed:
feat(lenel/openaccess): reader searching, raw lookup, event pagination (#403)

* feat(lenel/openaccess): reader searching, raw lookup, ignore alarm field

* style: lint, remove lenel comment

stakach committed:
feat(place/visitor_mailer): add option to disable qr

stakach committed:
chore: refactor router

more like a minor shuffling so things are cleaner

camreeves committed:
chore(leviton): Set debugging to info level to persist in logs (#407)

* chore(panel ui): add manual build of panel ui

* chore(leviton): minor cleanup and crystal tool format

stakach committed:
feat(cisco/dna_spaces): add sensor interface

stakach committed:
feat(place/bookings): allow filtering sensors by mac

where the room sensor is part of a larger collection

stakach committed:
fix(cisco/dna_spaces): sensor queries

stakach committed:
fix(cisco/dna_spaces): add units to sensor data

stakach committed:
feat(place/area_management): support transforming sensor units

for simple analytics and display

stakach committed:
fix(place/bookings): sensor data could be Int in older crystal versions

chillfox committed:
feat(survey): send invitation emails (#399)

stakach committed:
feat(cisco/meraki): add support for wired desk tracking (#412)

w-le committed:
feat(placeos/mailers): ISE network creds in Event and Booking emails (#405)

* feat(placeos/booking_notifier): support network credentials

* style: lint

* feat(placeos/mailers): ISE wifi creds in Event and Booking emails

* feat(cisco/ise): crated user's username is their email, instead of random

* feat(placeos/booking_notifier): simpler user password, update it each booking

* fix(placeos/booking_notifier): rescue syntax in network creds integration

* fix(placeos/booking_notifier): response handling of ISE integration

* fix(placeos/booking_notifier): response handling of ISE integration

* chore(cisco/ise): non random InternalUser name

* fix(cisco/ise): optional params for create_internal_user

* fix(cisco/ise): allow InternalUser initialization without name

* fix(cisco/ise): don't allow InternalUser initialization without name

* fix(cisco/ise): don't allow InternalUser initialization without name

* fix(placeos/event_mailer): compiles. untested.

* chore(placeos/event_mailer): generic_name

* fix(placeos/event_mailer): string matching

* fix(cisco/ise): random passwords are base64

* chore(placeos/event_mailer): debug output

* chore(placeos/event_mailer): debug output

* fix(cisco/ise): syntax

* fix(cisco/ise): syntax

* fix(cisco/ise): don't determine already processed events for now (always emails for 0 events for some reason)

* fix(placeos/event_mailer): syntax

* feat(placeos/visitor_mailer): support cisco ise network creds

* chore(cisco/ise): tidier json param formation

* style(cisco/ise): crystal tool format

* style(cisco/ise): lint

stakach committed:
fix(meraki_locations): mac can be nil on a port

stakach committed:
fix(meraki_locations): return desk locations using building zone

stakach committed:
fix(meraki_locations): ensure level_id is set on port status

stakach committed:
feat(meraki_locations): notify area management on port status changes

improves the speed of feedback getting to the users eyes

chillfox committed:
feat(survey_mailer): provide survey_id to template (#413)

stakach committed:
feat(event_attendance_recorder): at a debounce setting

for changes in sensor values

stakach committed:
chore: crystal tool format

w-le committed:
fix(placeos/booking_notifier): optional polling. rm unneeded wifi cred generation. (#414)

* fix(placeos/booking_notifier): polling now optional. fix wifi unneeded wifi cred generation.

* style(placeos/booking_notifier): lint

w-le committed:
fix(cisco/ise): reference user by name instead of email (#415)

grkek committed:
feat: Add identity groups to creation/update methods of the internal users (#416)

* Add identity groups to creation/update methods of the internal users

* Remove Kappture commit

* Add update user identity groups by name

* Move from 2 function calls to 1 function call

w-le committed:
feat(placeos/mailers): add groups for created network users

w-le committed:
fix(placeos/visitor_mailer): timezone setting

stakach committed:
feat(crestron/nvx_tx): expose the audio stream session name

stakach committed:
feat(crestron/virtual_switcher): add audio switching layer

allows simplified mixer config

stakach committed:
fix(crestron/virtual_switcher): lookup audio layer as required

stakach committed:
fix(webex/instant_connect): update driver endpoints

stakach committed:
fix(webex/instant_connect): revert API call changes

stakach committed:
fix(webex/instant_connect): be flexible about payload content

w-le committed:
fix(placeos/mailers): password for new network users

w-le committed:
Merge branch 'master' of github.com:PlaceOS/drivers

chillfox committed:
feat(visitor_mailer): multiple templates (#417)

stakach committed:
fix(cisco/dna_spaces): improve location search when no map available

stakach committed:
fix(event_attendance_recorder): delay initial value

also only record changes in people counts

stakach committed:
feat(place/bookings): add options to hide details

camreeves committed:
fix(leviton): allow for gzip decompression in log files (#411)

stakach committed:
feat(qsc/q_sys_control): add support for percentage based faders

stakach committed:
feat(qsc/q_sys_control): expose history data

stakach committed:
fix(place/staff_api): add nil defaults to get_survey_invites

chillfox committed:
fix(survey_mailer): update sent status (#420)

chillfox committed:
fix(staff_api): check sent for not nil in #get_survey_invites (#421)

stakach committed:
fix(event_attendance_recorder): ensure counts reset if metadata save fails

stakach committed:
chore(shard.lock): bump place driver

stakach committed:
chore(shard.lock): bump driver version

stakach committed:
chore(shard.lock): bump versions

stakach committed:
chore(cisco/dna_spaces): update default values

location age is a less useful metric as device exit events are used instead of constantly sending micro adjustments to location

stakach committed:
fix(cisco/dna_spaces): map_id is used for grouping results

so if there are two maps for a single level and they don't have a map id set, we need to use something else to group by

stakach committed:
feat: add auto desk checkin helper

camreeves committed:
fix(leviton): ensure right file is unzipped (#425)

stakach committed:
feat(area_management): add area capacity to areas

camreeves committed:
fix(leviton): update data format (#426)

stakach committed:
fix(acquisuite): rstrip new line chars from result

camreeves committed:
fix(leviton): update webhook logging to warn and output b64 before decoding (#429)

* fix(leviton): update webhook logging to warn and output b64 before decoding

* fix(leviton): undo shit code

* fix(leviton): undo shit code

stakach committed:
feat(bookings): expose some fields for analytics

stakach committed:
fix(bookings): all_day_event flag reversed

stakach committed:
feat(desk_bookings_locations): add duration for analytics

stakach committed:
feat(area_management): add a summary of desk bookings

as well as sensor detections

chillfox committed:
test(staff_api): #get_survey_invites (#430)

* test(staff_api): test #get_survey_invites

* test(staff_api): test #get_survey_invites

* style: format

chillfox committed:
feat(survey): dedupe emails (#432)

stakach committed:
feat(acquisuite): add timestamp hint

stakach committed:
feat(webex/instant_connect): add webex guest JWT

naqvis committed:
fix(docker-compose): update redis and install-shards image to multi-arch images tag (#433)

stakach committed:
feat: add health chat rooms (#431)

chillfox committed:
feat(visitor_mailor): all day event time PPT-107 (#424)

* fix(staff_api): check sent for not nil in #get_survey_invites

* feat(visitor_mailer): all day event time

* refactor(visitor_mailer)

* refactor(visitor_mailer)

stakach committed:
fix(chat/health): don't raise if not recently disconnected

stakach committed:
feat(leviton/acquisuite): persist last error to help with debugging

chillfox committed:
feat(password_generator): [PPT-90] add configurable password generator (#434)

stakach committed:
fix(crestron/nvx_rx): don't downcase stream names

stakach committed:
fix(crestron/nvx_rx): compilation issue

stakach committed:
feat(chat/health_rooms): add room summaries

for both the UI and analytics

stakach committed:
fix(chat/health_rooms): room summary wait time

should be the last time the user was updated

stakach committed:
fix(chat/health_rooms): wait time update check

stakach committed:
feat(chat/health_rooms_models): add text chat only flag to participants

stakach committed:
feat: add samsung reduced mdc protocol display

a display with a simplified control set

stakach committed:
feat(samsung/reduced_mdc_protocol): add whiteboard clear function

as I assume users can draw on the screen and the only way to clear it is to switch inputs and then switch back

stakach committed:
feat(chat/health_rooms): implement user exit signal

stakach committed:
feat(harness): ensure code & images are up to date

stakach committed:
fix(meraki/dashboard): webhook response code

stakach committed:
feat(place/area_management): obtain building id from system config

stakach committed:
fix(chat/health_rooms): ensure users are cleaned up

stakach committed:
feat(bookings): add ended at field for simplified analytics

stakach committed:
feat(chat/health_rooms): provide empty arrays when chat ends

to help with frontend updates

camreeves committed:
fix(leviton): ensure config list is loaded from settings (#435)

stakach committed:
feat: add johnson_controls metasys

w-le committed:
fix(place/event_mailer): saving of state event metadata (#436)

stakach committed:
feat(crestron/cres_next): added WS manual send function

for testing

stakach committed:
Merge branch 'master' of https://github.com/PlaceOS/drivers

stakach committed:
fix(crestron\nvx_rx): use websocket update method

as HTTP requests receiving 500 errors

stakach committed:
feat(crestron/virtual_switcher): add support for muting outputs

stakach committed:
revert(cisco/ise/guest): left for client compatibility

stakach committed:
chore(cisco/ise/guest): rename to avoid history clash

w-le committed:
feat: add globalcache ir (#437)

camreeves committed:
fix(levtion): allow for null values in readings (#438)

camreeves committed:
fix(leviton): Allow for readings to use the correct datatype (#439)

w-le committed:
fix(globalcache/ir_tv): add a delay between IR commands

w-le committed:
chore(globalcache/ir_tv): expose ir commands

stakach committed:
feat(place/bookings): add host_override option

also remove duplicate function in calendar common

stakach committed:
feat(place/staff-api): add metadata query function

stakach committed:
feat(staff_api): add get_event function

stakach committed:
fix(place/calendar): rate limiter fiber exit

also disable the rate limiter for graph API

stakach committed:
chore: fix generation of crystal docs

stakach committed:
limit meetings to 4 hours of age

stakach committed:
fix(health_rooms): disconnection validation check

reduce default timeouts

stakach committed:
feat(chat/health_rooms): save and restore state on update / recovery

stakach committed:
fix(place/meet): lighting independent setting

stakach committed:
chore: bump driver lighting interface

stakach committed:
feat(philips/dynet_text): add generic data send

stakach committed:
feat(philips/dynalite): add link_area function

stakach committed:
fix(philips\dynalite): fix join byte

stakach committed:
feat(philips/dynalite): improve unlink area

and limit area types to bytes

stakach committed:
feat(place/meet): improve light linking

stakach committed:
fix(place/bookings): potential for inaccurate status flag

stakach committed:
feat(place/staff_api): add ical_uid param to metadata patching

w-le committed:
fix(infosilem/room_schedule): current_event_attendees = 0 instead of nil when no event

grkek committed:
Add Vecos driver PPT-73 (#440)

grkek committed:
Add GoBright driver PPT-35 (#428)

stakach committed:
feat(gobright): simplify driver

stakach committed:
chore(shard.lock): bump versions

stakach committed:
fix(gobright/api): invalidate token if unauthorised response

stakach committed:
fix(gobright/api): spaces request can support multiple types

stakach committed:
fix(gobright/api): don't filter types by default

stakach committed:
feat: add GoBright location service

stakach committed:
fix(gobright/api): spaces request to return space object

stakach committed:
fix(gobright/api): live occupancy requires a location id

stakach committed:
fix(gobright/locations): return unoccupied spaces

when configured

stakach committed:
feat(place/chat): add entry notifications (#442)

stakach committed:
chore(health_rooms): add debugging log output

stakach committed:
chore(health_rooms): add debugging helper

stakach committed:
chore(health_rooms): expose internals in debugging helper

stakach committed:
fix(health_rooms): notification metadata extraction

stakach committed:
chore(health_rooms): load notification settings when inspecting

in case they are not set yet (as they are lazily loaded)

stakach committed:
feat(gobright/api): add bookings

stakach committed:
feat(gobright/api): allow data in response to be null

stakach committed:
fix(gobright/api): response might return data or items

stakach committed:
feat(gobright/location): include bookings in occupied status

stakach committed:
fix(releezme/vecos): don't require an external library

stakach committed:
feat(releezme/vecos): implement required sections of the API

stakach committed:
feat(bookings): add disable_book_now_host binding

stakach committed:
chore: rename releezme to vecos

stakach committed:
chore(vecos): crystal tool format

stakach committed:
feat(vecos/releezme): fix nullable fields

stakach committed:
fix(gobright): not all spaces have a location id

stakach committed:
feat(vecos/releezme): implement lockers interface

stakach committed:
fix(vecos/releezme): update models

stakach committed:
feat(vecos/releezme): implement location services

stakach committed:
fix(vecos\releezme_locations): update floor mappings key

stakach committed:
fix: unique keys for location services

stakach committed:
fix(vecos\releezme_locations): change floor mapping key

for more accuracy to the id being used

stakach committed:
chore(shard.lock): bump version

stakach committed:
feat(vecos/releezme_locations): skip fetching user if using email as id

stakach committed:
fix(vecos/releezme_locations): skip fetching user if using email as id

stakach committed:
chore: update lockers interface

stakach committed:
feat: add open ai chatgpt support

stakach committed:
fix(open_ai/gpt): listing available models

stakach committed:
feat(open_ai/gpt): make org optional and record usage stats

also rename module LLM as we can just use this as the basis for all LLM interfaces

stakach committed:
feat: add voice control template for LLM usage

stakach committed:
feat(open_ai/voice_control): provide access to status data

stakach committed:
fix(place/meet): default mute feedback is audio_mute

stakach committed:
chore(shard.lock): bump driver version

stakach committed:
fix(nec/display): resolve some common issues

audio input query is unsupported by many models
the checksum is incorrect for some firmware versions causing command retries

stakach committed:
chore(nec/display): add some debug logging

stakach committed:
fix(nec/display): multiple lengths for save response

grkek committed:
Add 25Live Pro driver PPT-106 (#441)

* Add twenty five live pro

* Added 25Live Pro API wrapper

* Add tests for the driver and add a room scheduler

* Run the formatter

* Remove the false data

* Swap out the parameter builder for the STD alternative

stakach committed:
chore(shard.lock): bump driver version

stakach committed:
fix(place/desk_bookings_locations): filter on location type

when provided

w-le committed:
chore(placeos/booking_checkin_helper): clearer func names for auto cancellation/prompting

stakach committed:
fix(floorsense/desks_websocket): poll to detect dead socket

stakach committed:
feat(nec/display): fix volume and mute feedback

w-le committed:
chore(place/booking_checin_helper): comment only

stakach committed:
feat(gobright/location_service): add support for desk bookings

stakach committed:
chore(shard.lock): bump versions

fixes issue with redis pool

grkek committed:
Bump items per page from 10 to 100 for 25 Live Pro API (#443)

grkek committed:
Finish Delta driver (#444)

stakach committed:
feat: add zencontrol classic tpi

stakach committed:
feat: add zencontrol using advanced TPI

stakach committed:
chore(shard.lock): bump versions

stakach committed:
feat(place/bookings): custom config for exposing data

stakach committed:
feat(desk_bookings_locations): can add custom keys

for analytics

stakach committed:
chore(place\bookings): fix code comment

stakach committed:
fix(lenel): may not have an email

camreeves committed:
fix(leviton): add SUCCESS to body of responses (#445)

Co-authored-by: Cameron Reeves <reeves.cameron@gmail.com>

stakach committed:
fix(amber_tech): can return a N/A state value

we default this to stopped

stakach committed:
chore(shard.lock): bump deps

grkek committed:
Add the switcher

grkek committed:
Update Kramer, add switchable interface

jeremyw24 commented:
@stakach can you please review this?
